### ISSUE TITLE
ASP-based solver: KeyError on spec construction

### ISSUE THREAD

**Issue by alalazo** [LLNL/spack#32348]
### Steps to reproduce

The error has been reported on Slack by @blue42u (on buildcaches) and @adamjstewart (on a local store). The steps to reproduce on a buildcache are:
```console
$ spack mirror add tmp-pr s3://spack-binaries-prs/pr31598_hpctoolkit-stack
$ spack -d spec autoconf
```

### Error message

```
==> [2022-08-24-06:55:05.094657] Imported spec from built-in commands
==> [2022-08-24-06:55:05.095964] Imported spec from built-in commands
==> [2022-08-24-06:55:05.097379] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/config.yaml
==> [2022-08-24-06:55:05.114014] Reading config file /home/culpo/.spack/config.yaml
==> [2022-08-24-06:55:05.114366] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/bootstrap.yaml
==> [2022-08-24-06:55:05.123078] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/config.yaml
==> [2022-08-24-06:55:05.135667] Reading config file /home/culpo/.spack/config.yaml
==> [2022-08-24-06:55:05.136020] DATABASE LOCK TIMEOUT: 3s
==> [2022-08-24-06:55:05.136071] PACKAGE LOCK TIMEOUT: No timeout
==> [2022-08-24-06:55:05.136525] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/bootstrap.yaml
==> [2022-08-24-06:55:05.141792] DATABASE LOCK TIMEOUT: 3s
==> [2022-08-24-06:55:05.141832] PACKAGE LOCK TIMEOUT: No timeout
==> [2022-08-24-06:55:05.141882] [BOOTSTRAP CONFIG SCOPE] name=_builtin
==> [2022-08-24-06:55:05.142169] [BOOTSTRAP CONFIG SCOPE] name=defaults, path=/home/culpo/PycharmProjects/spack/etc/spack/defaults
==> [2022-08-24-06:55:05.142198] [BOOTSTRAP CONFIG SCOPE] name=defaults/linux, path=/home/culpo/PycharmProjects/spack/etc/spack/defaults/linux
==> [2022-08-24-06:55:05.142221] [BOOTSTRAP CONFIG SCOPE] name=bootstrap, path=/home/culpo/.spack/bootstrap/config
==> [2022-08-24-06:55:05.142245] [BOOTSTRAP CONFIG SCOPE] name=bootstrap/linux, path=/home/culpo/.spack/bootstrap/config/linux
==> [2022-08-24-06:55:05.142459] Reading config file /home/culpo/.spack/bootstrap/config/linux/compilers.yaml
==> [2022-08-24-06:55:05.176990] [BOOTSTRAP ROOT SPEC] clingo-bootstrap@spack+python %gcc target=x86_64
==> [2022-08-24-06:55:05.177070] [BOOTSTRAP MODULE clingo] Try importing from Python
==> [2022-08-24-06:55:05.177279] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/bootstrap.yaml
==> [2022-08-24-06:55:05.293748] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/config.yaml
==> [2022-08-24-06:55:05.345674] [BOOTSTRAP MODULE clingo] The installed spec "clingo-bootstrap@spack+python %gcc target=x86_64 ^python@3.8 /shqedxgvjnhiwdcdrvjhbd73jaevv7wt" provides the "clingo" Python module
==> [2022-08-24-06:55:05.345934] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/concretizer.yaml
==> [2022-08-24-06:55:05.349508] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/repos.yaml
==> [2022-08-24-06:55:05.350756] Reading config file /home/culpo/.spack/repos.yaml
==> [2022-08-24-06:55:05.374774] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/config.yaml
==> [2022-08-24-06:55:05.387145] Reading config file /home/culpo/.spack/config.yaml
==> [2022-08-24-06:55:05.387763] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/mirrors.yaml
==> [2022-08-24-06:55:05.388972] Reading config file /home/culpo/.spack/mirrors.yaml
==> [2022-08-24-06:55:10.157338] Cached index for s3://spack-binaries-prs/pr31598_hpctoolkit-stack already up to date
==> [2022-08-24-06:55:10.891502] DATABASE LOCK TIMEOUT: 3s
==> [2022-08-24-06:55:10.891628] PACKAGE LOCK TIMEOUT: No timeout
==> [2022-08-24-06:55:12.690787] Reading config file /home/culpo/.spack/linux/compilers.yaml
==> [2022-08-24-06:55:12.715401] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml
Traceback (most recent call last):
  File "/home/culpo/PycharmProjects/spack/bin/spack", line 100, in <module>
    sys.exit(spack.main.main())
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 961, in main
    return _main(argv)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 916, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 944, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 608, in _invoke_command
    return_val = command(parser, args)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/cmd/spec.py", line 104, in spec
    concretized_specs = spack.cmd.parse_specs(args.specs, concretize=True)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/cmd/__init__.py", line 227, in parse_specs
    spec.concretize(tests=tests)  # implies normalize
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/spec.py", line 2924, in concretize
    self._new_concretize(tests)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/spec.py", line 2899, in _new_concretize
    result = solver.solve([self], tests=tests)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/solver/asp.py", line 2318, in solve
    result, _, _ = self.driver.solve(setup, specs, reuse=reusable_specs, output=output)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/solver/asp.py", line 695, in solve
    answers = builder.build_specs(tuples)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/solver/asp.py", line 2177, in build_specs
    action(*args)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/solver/asp.py", line 1984, in hash
    self._specs[pkg] = self._hash_lookup[h]
KeyError: 'tg6ojo7ktvgvnlna557wf7sbdg4zz45x'
```

### Information on your system

* **Spack:** 0.19.0.dev0 (af4788fdefd156efdb9ce7ffbeb57c8f73898e43)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by alalazo** [comment]
The issue seems to be a missing hash for a transitive dependency. In the ASP code I see:
```
installed_hash("perl","ff4kmp54dj4thbwawypsytvpq2wk2cry").
imposed_constraint("ff4kmp54dj4thbwawypsytvpq2wk2cry","node","perl").
imposed_constraint("ff4kmp54dj4thbwawypsytvpq2wk2cry","version","perl","5.34.1").
imposed_constraint("ff4kmp54dj4thbwawypsytvpq2wk2cry","node_platform","perl","linux").
imposed_constraint("ff4kmp54dj4thbwawypsytvpq2wk2cry","node_os","perl","ubuntu20.04").
imposed_constraint("ff4kmp54dj4thbwawypsytvpq2wk2cry","node_target","perl","x86_64_v3").
imposed_constraint("ff4kmp54dj4thbwawypsytvpq2wk2cry","variant_value","perl","cpanm","True").
imposed_constraint("ff4kmp54dj4thbwawypsytvpq2wk2cry","variant_value","perl","shared","True").
imposed_constraint("ff4kmp54dj4thbwawypsytvpq2wk2cry","variant_value","perl","threads","True").
imposed_constraint("ff4kmp54dj4thbwawypsytvpq2wk2cry","node_compiler","perl","gcc").
imposed_constraint("ff4kmp54dj4thbwawypsytvpq2wk2cry","node_compiler_version","perl","gcc","9.4.0").
imposed_constraint("ff4kmp54dj4thbwawypsytvpq2wk2cry","hash","perl","ff4kmp54dj4thbwawypsytvpq2wk2cry").
imposed_constraint("ff4kmp54dj4thbwawypsytvpq2wk2cry","depends_on","perl","berkeley-db","link").
imposed_constraint("ff4kmp54dj4thbwawypsytvpq2wk2cry","hash","berkeley-db","hirmxka5mkpr22hh4cr6pvztvwj6wpi2").
imposed_constraint("ff4kmp54dj4thbwawypsytvpq2wk2cry","depends_on","perl","bzip2","link").
imposed_constraint("ff4kmp54dj4thbwawypsytvpq2wk2cry","hash","bzip2","oobld7xhovhcgx6nrwujeyocl62j437y").
imposed_constraint("ff4kmp54dj4thbwawypsytvpq2wk2cry","depends_on","perl","gdbm","link").
imposed_constraint("ff4kmp54dj4thbwawypsytvpq2wk2cry","hash","gdbm","tg6ojo7ktvgvnlna557wf7sbdg4zz45x").
imposed_constraint("ff4kmp54dj4thbwawypsytvpq2wk2cry","depends_on","perl","zlib","link").
imposed_constraint("ff4kmp54dj4thbwawypsytvpq2wk2cry","hash","zlib","ckadtmkfuxuyizgp5djpvv7felahwrcf").
```
and the `hash` in the `KeyError` refers to `gdbm`. In the logic program though, there is no `installed_hash` fact emitted for it. Investigating now why the hash is missing.

**Comment by alalazo** [comment]
So this issue started after #30806 but I think the PR just made a latent bug apparent. In the index we miss the hash of a few dependencies so, for instance, for the `autoconf` above we miss `gdbm`, `readline` and `ncurses`. 

If we revert #30806 the behavior is to select a buildcache, and then install the missing spec from sources. Given that we might avoid default values to select a buildcache because we can "reuse" this seems too a faulty behavior - i.e. if we can't install from binaries we should just use the defaults for variants versions etc.

After #30806 instead we assume that a binary mirror is self-consistent - if it has a spec for `perl` it should also conatin all of its transitive dependencies. This is where we are failing now.

**Comment by alalazo** [comment]
In a case like this (hash `/xyz` is in the buildcache, but one or more of of its dependencies is not) how do we want Spack to behave? Should it:
1. Consider the hash in the solve, and go back to the previous behavior of building missing dependencies, or
2. Discard that hash and emit a warning or a debug level message

? I would lean towards 2.

**Comment by alalazo** [comment]
Still trying to analyze the root cause of this, since I can't reproduce the error by creating syntetic buildcaches with a missing dependency.

**Comment by alalazo** [comment]
I think this might be the right way to fix the issue:
```diff
diff --git a/lib/spack/spack/solver/concretize.lp b/lib/spack/spack/solver/concretize.lp
index 48cb0f1d36..6cc2ddcdb1 100644
--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -1059,6 +1059,9 @@ no_flags(Package, FlagType)
 % you can't choose an installed hash for a dev spec
 :- hash(Package, Hash), variant_value(Package, "dev_path", _).
 
+% You can't install a hash, if it is not installed
+:- hash(Package, Hash), not installed_hash(Package, Hash).
+
 % if a hash is selected, we impose all the constraints that implies
 impose(Hash) :- hash(Package, Hash).
```
I tried that in @blue42u use case and it works. I'm still trying to understand what chain of deduction can lead to have the hash of `gdbm` as an atom in the answer set (i.e. I'm trying to figure out if this integrity constraint, which is legit, covers up for a bug lurking in the logic program).

**Comment by alalazo** [comment]
It seems simple synthetic cases would hit different errors if we try to force the hash of a package with a missing dependency. For instance, trying to concretize `libpng`:
```console
Input spec
--------------------------------
libpng

Concretized
--------------------------------
libpng@1.6.37%gcc@12.1.0 arch=linux-ubuntu20.04-icelake
    ^zlib@1.2.12%gcc@12.1.0+optimize+pic+shared patches=0d38234 arch=linux-ubuntu20.04-icelake
```
from a buildcache with only `libpng` and without `zlib` leads to errors like:
```
error(2,"No valid compiler version found for '{0}'","zlib")
```
if we force clingo to pick `libpng` hash. For more complicated cases, where we have a lot more specs in the buildcache, we don't have logic preventing:
1. The presence of two different hashes from the same package, if they are compatible 
2. The presence of specs that have dependencies missing in the buildcache

In the case of @blue42u forcing the presence of `tg6ojo7ktvgvnlna557wf7sbdg4zz45x` I obtained this answer set:
```
impose("tg6ojo7ktvgvnlna557wf7sbdg4zz45x") hash("gdbm","tg6ojo7ktvgvnlna557wf7sbdg4zz45x") impose(1) hash("autoconf","hw3jnkopzspnrqnxl5cclvwloln52ife") impose("hw3jnkopzspnrqnxl5cclvwloln52ife") hash("perl","ff4kmp54dj4thbwawypsytvpq2wk2cry") hash("m4","lh54wz2itpbjhkjp3apbh5dxcdj7yawn") impose("lh54wz2itpbjhkjp3apbh5dxcdj7yawn") impose("ff4kmp54dj4thbwawypsytvpq2wk2cry") hash("libsigsegv","qre4xmp3zva33bwi34nwavtanrfnwgxe") hash("zlib","ckadtmkfuxuyizgp5djpvv7felahwrcf") hash("gdbm","qj255brd2jeniylvy32632esn6iog2pn") hash("bzip2","oobld7xhovhcgx6nrwujeyocl62j437y") hash("berkeley-db","hirmxka5mkpr22hh4cr6pvztvwj6wpi2") impose("hirmxka5mkpr22hh4cr6pvztvwj6wpi2") impose("oobld7xhovhcgx6nrwujeyocl62j437y") impose("qj255brd2jeniylvy32632esn6iog2pn") impose("ckadtmkfuxuyizgp5djpvv7felahwrcf") impose("qre4xmp3zva33bwi34nwavtanrfnwgxe") hash("readline","crkj7ff4vvlutccd2t24fbrob6m32ldo") impose("crkj7ff4vvlutccd2t24fbrob6m32ldo") hash("ncurses","3zo3eo3zshu74qy3id32fmvopb2yq63c") impose("3zo3eo3zshu74qy3id32fmvopb2yq63c") impose(8) impose(11) impose(13) impose(28) impose(44) impose(49) impose(56) impose(63) impose(65) impose(68) impose(70) impose(94)
```
where `gdbm` is present with 2 hashes. This is because the only constraint we have on it is conditional on the hash being one among the installed hashes:
```
{ hash(Package, Hash) : installed_hash(Package, Hash) } 1
 :- node(Package), internal_error("Package must resolve to at most one hash").
```
which is respected since the hash forced is missing in the buildcache.

**Comment by blue42u** [comment]
Thanks! I managed to throw together a rather large reproducer based on your analysis and what I know from #32193, I tried a few smaller examples but this was the only one that crashed:
<details>

```console
$ cat x/spack.yaml 
spack:
  specs:
  - autoconf
$ spack -e x install
==> Starting concretization
==> Environment concretized in 0.72 seconds.
==> Concretized autoconf
 -   tpgif6m  autoconf@2.69%gcc@9.4.0 patches=35c4492,7793209,a49dd5b arch=linux-ubuntu20.04-zen2
 -   lmp3mwi      ^m4@1.4.19%gcc@9.4.0+sigsegv patches=9dc5fbd,bfdffa7 arch=linux-ubuntu20.04-zen2
 -   fnufbju          ^diffutils@3.8%gcc@9.4.0 arch=linux-ubuntu20.04-zen2
 -   xmcs67a              ^libiconv@1.16%gcc@9.4.0 libs=shared,static arch=linux-ubuntu20.04-zen2
 -   azfnq5a          ^libsigsegv@2.13%gcc@9.4.0 arch=linux-ubuntu20.04-zen2
 -   mynxfdc      ^perl@5.34.1%gcc@9.4.0+cpanm+shared+threads arch=linux-ubuntu20.04-zen2
 -   m5ephzc          ^berkeley-db@18.1.40%gcc@9.4.0+cxx~docs+stl patches=b231fcc arch=linux-ubuntu20.04-zen2
 -   i3gd6bh          ^bzip2@1.0.8%gcc@9.4.0~debug~pic+shared arch=linux-ubuntu20.04-zen2
 -   zmkskq2          ^gdbm@1.19%gcc@9.4.0 arch=linux-ubuntu20.04-zen2
 -   zniuxhi              ^readline@8.1.2%gcc@9.4.0 arch=linux-ubuntu20.04-zen2
 -   5fq7pwt                  ^ncurses@6.3%gcc@9.4.0~symlinks+termlib abi=none arch=linux-ubuntu20.04-zen2
 -   izrtaeu                      ^pkgconf@1.8.0%gcc@9.4.0 arch=linux-ubuntu20.04-zen2
 -   jbd75by          ^zlib@1.2.12%gcc@9.4.0+optimize+pic+shared patches=0d38234 arch=linux-ubuntu20.04-zen2
...
$ spack -e x buildcache create -aud /tmp/xy_mirror
==> Pushing binary packages to file:///tmp/xy_mirror/build_cache
$ rm -v /tmp/xy_mirror/build_cache/*-{gdbm,ncurses,readline}-*.spec.json 
removed '/tmp/xy_mirror/build_cache/linux-ubuntu20.04-zen2-gcc-9.4.0-gdbm-1.19-zmkskq222hsrz77ojakne3yaku2jifyo.spec.json'
removed '/tmp/xy_mirror/build_cache/linux-ubuntu20.04-zen2-gcc-9.4.0-ncurses-6.3-5fq7pwtlfrcdilllvrnoaqyte4lw2rgr.spec.json'
removed '/tmp/xy_mirror/build_cache/linux-ubuntu20.04-zen2-gcc-9.4.0-readline-8.1.2-zniuxhiyn7qobi2sfat4clcfg5jimjn5.spec.json'
$ spack uninstall -ay
...

$ patch -p1 -d /opt/spack <<'EOF'
diff --git a/lib/spack/spack/spec.py b/lib/spack/spack/spec.py
index 8a463c2f78..36beffd5a5 100644
--- a/lib/spack/spack/spec.py
+++ b/lib/spack/spack/spec.py
@@ -1788,6 +1788,7 @@ def spec_hash(self, hash):
             return hash.override(self)
         node_dict = self.to_node_dict(hash=hash)
         json_text = sjson.dump(node_dict)
+        json_text += 'CHANGE!'
         return spack.util.hash.b32_hash(json_text)
 
     def _cached_hash(self, hash, length=None, force=False):
EOF
patching file lib/spack/spack/spec.py
$ cat y/spack.yaml
spack:
  specs:
  - gdbm
$ spack -e y install
==> Starting concretization
==> Environment concretized in 0.26 seconds.
==> Concretized gdbm
 -   5vne3vx  gdbm@1.23%gcc@9.4.0 arch=linux-ubuntu20.04-zen2
 -   lfjtext      ^readline@8.1.2%gcc@9.4.0 arch=linux-ubuntu20.04-zen2
 -   hzhsyad          ^ncurses@6.3%gcc@9.4.0~symlinks+termlib abi=none arch=linux-ubuntu20.04-zen2
 -   mf7m63t              ^pkgconf@1.8.0%gcc@9.4.0 arch=linux-ubuntu20.04-zen2
...
$ spack -e y buildcache create -aud /tmp/xy_mirror
==> Pushing binary packages to file:///tmp/xy_mirror/build_cache
$ spack uninstall -ay
...

$ spack buildcache update-index -d /tmp/xy_mirror
$ cat z/spack.yaml 
spack:
  mirrors:
    xy_mirror: /tmp/xy_mirror
$ spack -d -e z spec autoconf
==> [2022-08-24-15:20:08.402051] Reading config file /opt/spack/etc/spack/defaults/config.yaml
==> [2022-08-24-15:20:08.415573] Reading config file /opt/spack/etc/spack/defaults/concretizer.yaml
==> [2022-08-24-15:20:08.419451] Reading config file /root/z/spack.yaml
==> [2022-08-24-15:20:08.420506] Using environment '/root/z'
==> [2022-08-24-15:20:08.421179] Imported spec from built-in commands
==> [2022-08-24-15:20:08.421614] Imported spec from built-in commands
==> [2022-08-24-15:20:08.422066] Reading config file /opt/spack/etc/spack/defaults/bootstrap.yaml
==> [2022-08-24-15:20:08.426026] Deactivated environment '/root/z'
==> [2022-08-24-15:20:08.445127] Reading config file /opt/spack/etc/spack/defaults/config.yaml
==> [2022-08-24-15:20:08.456440] DATABASE LOCK TIMEOUT: 3s
==> [2022-08-24-15:20:08.456471] PACKAGE LOCK TIMEOUT: No timeout
==> [2022-08-24-15:20:08.456711] Reading config file /opt/spack/etc/spack/defaults/bootstrap.yaml
==> [2022-08-24-15:20:08.460425] DATABASE LOCK TIMEOUT: 3s
==> [2022-08-24-15:20:08.460453] PACKAGE LOCK TIMEOUT: No timeout
==> [2022-08-24-15:20:08.460489] [BOOTSTRAP CONFIG SCOPE] name=_builtin
==> [2022-08-24-15:20:08.460728] [BOOTSTRAP CONFIG SCOPE] name=defaults, path=/opt/spack/etc/spack/defaults
==> [2022-08-24-15:20:08.460752] [BOOTSTRAP CONFIG SCOPE] name=defaults/linux, path=/opt/spack/etc/spack/defaults/linux
==> [2022-08-24-15:20:08.460772] [BOOTSTRAP CONFIG SCOPE] name=bootstrap, path=/root/.spack/bootstrap/config
==> [2022-08-24-15:20:08.460787] [BOOTSTRAP CONFIG SCOPE] name=bootstrap/linux, path=/root/.spack/bootstrap/config/linux
==> [2022-08-24-15:20:08.460964] Reading config file /root/.spack/bootstrap/config/linux/compilers.yaml
==> [2022-08-24-15:20:08.475373] [BOOTSTRAP ROOT SPEC] clingo-bootstrap@spack+python %gcc target=x86_64
==> [2022-08-24-15:20:08.475428] [BOOTSTRAP MODULE clingo] Try importing from Python
==> [2022-08-24-15:20:08.475571] Reading config file /opt/spack/etc/spack/defaults/bootstrap.yaml
==> [2022-08-24-15:20:08.587826] Reading config file /opt/spack/etc/spack/defaults/config.yaml
==> [2022-08-24-15:20:08.632332] [BOOTSTRAP MODULE clingo] The installed spec "clingo-bootstrap@spack+python %gcc target=x86_64 ^python@3.8 /shqedxgvjnhiwdcdrvjhbd73jaevv7wt" provides the "clingo" Python module
==> [2022-08-24-15:20:08.632866] Reading config file /opt/spack/etc/spack/defaults/config.yaml
==> [2022-08-24-15:20:08.645074] Reading config file /root/z/spack.yaml
==> [2022-08-24-15:20:08.645854] Using environment '/root/z'
==> [2022-08-24-15:20:08.645922] Reading config file /opt/spack/etc/spack/defaults/concretizer.yaml
==> [2022-08-24-15:20:08.649203] Reading config file /opt/spack/etc/spack/defaults/repos.yaml
==> [2022-08-24-15:20:08.655021] Reading config file /opt/spack/etc/spack/defaults/mirrors.yaml
==> [2022-08-24-15:20:08.658685] Cached index for /tmp/xy_mirror already up to date
==> [2022-08-24-15:20:08.812845] DATABASE LOCK TIMEOUT: 3s
==> [2022-08-24-15:20:08.812892] PACKAGE LOCK TIMEOUT: No timeout
==> [2022-08-24-15:20:08.858045] Reading config file /root/.spack/linux/compilers.yaml
==> [2022-08-24-15:20:08.861856] Reading config file /opt/spack/etc/spack/defaults/packages.yaml
Traceback (most recent call last):
  File "/opt/spack/bin/spack", line 100, in <module>
    sys.exit(spack.main.main())
  File "/opt/spack/lib/spack/spack/main.py", line 961, in main
    return _main(argv)
  File "/opt/spack/lib/spack/spack/main.py", line 916, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/opt/spack/lib/spack/spack/main.py", line 944, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/opt/spack/lib/spack/spack/main.py", line 608, in _invoke_command
    return_val = command(parser, args)
  File "/opt/spack/lib/spack/spack/cmd/spec.py", line 104, in spec
    concretized_specs = spack.cmd.parse_specs(args.specs, concretize=True)
  File "/opt/spack/lib/spack/spack/cmd/__init__.py", line 227, in parse_specs
    spec.concretize(tests=tests)  # implies normalize
  File "/opt/spack/lib/spack/spack/spec.py", line 2925, in concretize
    self._new_concretize(tests)
  File "/opt/spack/lib/spack/spack/spec.py", line 2900, in _new_concretize
    result = solver.solve([self], tests=tests)
  File "/opt/spack/lib/spack/spack/solver/asp.py", line 2347, in solve
    result, _, _ = self.driver.solve(setup, specs, reuse=reusable_specs, output=output)
  File "/opt/spack/lib/spack/spack/solver/asp.py", line 706, in solve
    answers = builder.build_specs(tuples)
  File "/opt/spack/lib/spack/spack/solver/asp.py", line 2199, in build_specs
    action(*args)
  File "/opt/spack/lib/spack/spack/solver/asp.py", line 2002, in hash
    self._specs[pkg] = self._hash_lookup[h]
KeyError: 'zmkskq222hsrz77ojakne3yaku2jifyo'
```

</details>

I'm wondering if the root cause behind this and #32193 are the same, Spack trusts the `index.json` to contain a valid hash and copies it into the Spec on load: https://github.com/spack/spack/blob/ee02623171475df22be8d70ed2075cadff0c4a93/lib/spack/spack/spec.py#L2133-L2134
Then the concretizer assumes that every Spec has exactly one hash... which it might not if the hash changed recently and the `index.json` was out of date. And when that happens all hell breaks loose.

Would the right solution then be to re-hash Specs on load from JSON? That would ensure consistency without ASP magic.

**Comment by alalazo** [comment]
Can you check #32354? I think this solves the issue.

**Comment by alalazo** [comment]
> Then the concretizer assumes that every Spec has exactly one hash... which it might not if the hash changed recently and the index.json was out of date. And when that happens all hell breaks loose.

I think the problem is different. The buildcache contains a few binary packages that don't have the corresponding dependencies, and we were instead assuming that if an `installed_hash` was in a buildcache so were its dependencies. Adding an integrity constraint solves the issue in the logic program, because it makes the answer sets that lead to these errors not valid solutions anymore.

**Comment by tgamblin** [comment]
Is this just a documentation issue?  In the PR, the build cache is set up sort of like this:

```yaml
mirrors:
    tmp-pr: s3://spack-binaries-prs/pr31598_hpctoolkit-stack
    develop: s3://spack-binaries-prs/develop
    spack-public: https://mirror.spack.io
```

i.e., things not built specifically by the PR are fetched from the develop buildcache.  The PR build caches are fundamentally incomplete in that we only put things built *for* the PR in them, and some dependencies will necessarily come from `develop`

We could set up mirrors with some config to point to dependency mirrors, but we haven't really done that yet.

I think Max's fix in #32354 is actually good because it makes Spack handle this gracefully.  Also because the rolling `develop` build cache is quite large, and we likely want to start purging old builds from it.  #32354 might let us do that more easily using AWS's builtin object TTL policies instead of a lot of analysis,. but I haven't completely thought the through.

**Comment by blue42u** [comment]
@tgamblin It's more than a documentation issue if the PR builds something completely outside the public buildcache. I first saw this issue when I tried [pulling CentOS 7 binaries from the PR buildcache](https://gitlab.com/hpctoolkit/hpctoolkit/-/jobs/2908440433). :upside_down_face:

FYI `s3://spack-binaries-prs/pr31598_hpctoolkit-stack` + rolling `develop` buildcache worked for us until August 15th (exactly 1 month after I opened #31598). Then I started seeing this and #32193.

No objection to #32354, I would urge merging it ASAP.
